vertx-discovery-service/src/main/java/io/vertx/ext/discovery/DiscoveryService.java

```diff
@@ -0,0 +1,179 @@
+ * * un-publish a published record
+ * * update the status of a published service (down, out of service...)
+ * <p>
+ * A service consumer can:
+ * <p>
+ * * lookup for services
+ * * bind to a selected service (it gets a {@link ServiceReference}) and use it
+ * * release the service once the consumer is done with it
+ * * listen for arrival, departure and modification of services.
+ * <p>
+ * Consumer would 1) lookup for service record matching their need, 2) retrieve the {@link ServiceReference} that give access
+ * to the service, 3) get a service object to access the service, 4) release the service object once done.
+ * <p>
+ * A state above, the central piece of information shared by the providers and consumers are {@link Record records}.
+ * <p>
+ * Providers and consumers must create their own {@link DiscoveryService} instance. These instances are collaborating
+ * in background (distributed structure) to keep the set of services in sync.
+ *
+ * @author <a href="http://escoffier.me">Clement Escoffier</a>
+ */
+@VertxGen
+public interface DiscoveryService {
+
+  /**
+   * Creates an instance of {@link DiscoveryService}.
+   *
+   * @param vertx   the vert.x instance
+   * @param options the discovery options
+   * @return the create discovery service.
+   */
+  static DiscoveryService create(Vertx vertx, DiscoveryOptions options) {
+    return new DiscoveryImpl(vertx, options);
+  }
+
+  /**
+   * Creates a new instance of {@link DiscoveryService} using the default configuration.
+   *
+   * @param vertx the vert.x instance
+   * @return the created instance
+   */
+  static DiscoveryService create(Vertx vertx) {
+    return create(vertx, new DiscoveryOptions());
+  }
+
+  /**
+   * Gets a service from the selected record.
+   *
+   * @param vertx  the vert.x instance
+   * @param record the chosen record
+   * @return the service, that allows retrieving the service object
+   */
+  static ServiceReference getServiceReference(Vertx vertx, Record record) {
+    Objects.requireNonNull(record);
+    Objects.requireNonNull(vertx);
+    return ServiceTypes.get(record).get(vertx, record);
+  }
+
+  /**
+   * Registers a discovery bridge. Bridges let you integrate other discovery technologies in this discovery service.
+   *
+   * @param bridge        the bridge
+   * @param configuration the optional configuration
+   * @return the current {@link DiscoveryService}
+   */
+  @GenIgnore
+  DiscoveryService registerDiscoveryBridge(DiscoveryBridge bridge, JsonObject configuration);
+
+  /**
+   * Closes the discovery service
+   */
+  void close();
+
+  /**
+   * Publishes a record.
+   *
+   * @param record        the record
+   * @param resultHandler handler called when the operation has completed (successfully or not). In case of success,
+   *                      the passed record has a registration id required to modify and un-register the service.
+   */
+  void publish(Record record, Handler<AsyncResult<Record>> resultHandler);
```

Comment:
then add it to the doc :-)

> On Apr 1, 2016, at 12:14 PM, Clement Escoffier <notifications@github.com> wrote:
> 
> In vertx-discovery-service/src/main/java/io/vertx/ext/discovery/DiscoveryService.java <https://github.com/vert-x3/vertx-microservice-toolbox/pull/2#discussion_r58187018>:
> 
> > +  @GenIgnore
> > +  DiscoveryService registerDiscoveryBridge(DiscoveryBridge bridge, JsonObject configuration);
> > +
> > +  /**
> > +   * Closes the discovery service
> > +   */
> > +  void close();
> > +
> > +  /**
> > +   * Publishes a record.
> > +   *
> > +   * @param record        the record
> > +   * @param resultHandler handler called when the operation has completed (successfully or not). In case of success,
> > +   *                      the passed record has a registration id required to modify and un-register the service.
> > +   */
> > +  void publish(Record record, Handler<AsyncResult<Record>> resultHandler);
> Why ? The record can have been modified (it actually is modified, as the registration is set).
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub <https://github.com/vert-x3/vertx-microservice-toolbox/pull/2/files/34c3d23a0f85f72099c2095425457a1ab5a9412d#r58187018>